### PR TITLE
Deprecate namespace strategies other than 'per user'

### LINKED ISSUE #19365: Deprecate namespace strategies other than "per user"
### Is your task related to a problem? Please describe.

When switching to DevWorkspace we would like to drop support for namespace strategies other than "per user".
One and only one namespace per user simplifies a lot the design of the workspaces.
We could assume then that every user configuration lives in his namespace. 

### Describe the solution you'd like
At the "deprecation" stage we need to reflect that fact on documentation and add a warning when someone wants to use it.
 
### Describe alternatives you've considered
Remove support without "deprecation" stage

### Additional context
n/a

### PR DESCRIPTION
**Pull request by skabashnyuk**
### What does this PR do?
Deprecate namespace strategies other than 'per user'

### What issues does this PR fix or reference?
Fixes #19365 

### How to test this PR?
 - deploy che server image `quay.io/skabashn/che-server:che19365`
 - Setup namespace strategies other than 'per user'
 - you should see a warning message in the logs.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 1]
```diff
@@ -179,6 +187,36 @@ KubernetesNamespace doCreateNamespaceAccess(String workspaceId, String name) {
         clientFactory, cheClientFactory, sharedPool.getExecutor(), name, workspaceId);
   }
 
+  @ScheduleRate(period = 1, initialDelay = 1, unit = TimeUnit.HOURS)
```
Comment (mshaposhnik) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java:191]: imo printing warn on server startup is more efficient. Make sense consider changing delay to 0 ?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 2]
```diff
@@ -179,6 +187,36 @@ KubernetesNamespace doCreateNamespaceAccess(String workspaceId, String name) {
...
+  private void printLegacyWarningIfNeeded() {
+    if (!isNullOrEmpty(legacyNamespaceName)) {
+      LOG.warn(
+          "'che.infra.kubernetes.namespace' configuration parameter has been deprecated "
+              + "and is subject to remove in future releases.  Current value `{}`. Legacy workspaces located "
```
Comment (mshaposhnik) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: Current value `is` `{}`. \ Current values `are` `{}`.

**Review by mshaposhnik — APPROVED** [review]

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 3]
```diff
@@ -179,6 +187,36 @@ KubernetesNamespace doCreateNamespaceAccess(String workspaceId, String name) {
...
+              + "in this namespace may become unreachable in future releases. "
+              + "Please refer to the documentation about possible next steps.",
+          legacyNamespaceName);
+    }
+    if (LEGACY_NAMESPACE_NAME_PLACEHOLDERS.stream().anyMatch(defaultNamespaceName::contains)) {
+      LOG.warn(
+          "'che.infra.kubernetes.namespace.default' configuration parameter with `{}` "
+              + "placeholder has been deprecated and is subject to remove in  future releases."
+              + " Current value `{}`. Please refer to the documentation about "
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: Current value is `{}`.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 4]
```diff
@@ -179,6 +187,36 @@ KubernetesNamespace doCreateNamespaceAccess(String workspaceId, String name) {
...
+              + "possible next steps.",
+          LEGACY_NAMESPACE_NAME_PLACEHOLDERS,
+          defaultNamespaceName);
+    } else if (REQUIRED_NAMESPACE_NAME_PLACEHOLDERS
+        .stream()
+        .noneMatch(defaultNamespaceName::contains)) {
+      LOG.warn(
+          "Namespace strategies other than 'per user' has been deprecated and are subject to remove in  in future releases. "
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: ... have been deprecated and are subject to removal in future releases.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 5]
```diff
@@ -179,6 +187,36 @@ KubernetesNamespace doCreateNamespaceAccess(String workspaceId, String name) {
...
+              + "Usage of one placeholder from `{}` is required in 'che.infra.kubernetes.namespace.default' parameter."
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: Maybe something like "Using the <userid> or <username> placeholder is required ..."?

**Review by metlos — COMMENTED** [review]
LGTM with some minor grammar suggestions.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java (hunk 1 above)
Comment (skabashnyuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java:191]: this method is called in this class constructor. As a result of setting the delay to 0, we will see two identical messages almost at the same time.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 6]
```diff
@@ -311,7 +314,9 @@ che.infra.kubernetes.namespace.creation_allowed=true
 # if user does not override it.
 # It's possible to use <username>, <userid> and <workspaceid> placeholders (e.g.: che-workspace-<username>).
 # In that case, new namespace will be created for each user (or workspace).
-# Is used by OpenShift infra as well to specify Project
+# Is used by OpenShift infra as well to specify Project.
+# <workspaceid> placeholder is deprecated and is subject to removal in future releases
+# <username> or <userid> placeholders is mandatory. One and only one has to be part of the parameter.
```
[the remark this answers is not in the document]
Comment (sparkoo) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: why "one and only"? Having both of them is 'per user' as well. Also I believe this check is not implemented in `printLegacyWarningIfNeeded` method.

**Review by tolusha — APPROVED** [review]

**Review by sparkoo — APPROVED** [review]

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 7]
```diff
@@ -311,7 +314,9 @@ che.infra.kubernetes.namespace.creation_allowed=true
 # if user does not override it.
 # It's possible to use <username>, <userid> and <workspaceid> placeholders (e.g.: che-workspace-<username>).
 # In that case, new namespace will be created for each user (or workspace).
-# Is used by OpenShift infra as well to specify Project
+# Is used by OpenShift infra as well to specify Project.
```
Comment (rkratky) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]:
```suggestion
# Used by OpenShift infra as well to specify a Project.
```

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 8]
```diff
@@ -311,7 +314,9 @@ che.infra.kubernetes.namespace.creation_allowed=true
...
+# <workspaceid> placeholder is deprecated and is subject to removal in future releases
```
Comment (rkratky) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]:
```suggestion
# The `<workspaceid>` placeholder is deprecated and is subject to removal in future releases.
```

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 9]
```diff
@@ -311,7 +314,9 @@ che.infra.kubernetes.namespace.creation_allowed=true
...
+# <username> or <userid> placeholders is mandatory.
```
Comment (rkratky) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]:
```suggestion
# The `<username>` or `<userid>` placeholder is mandatory.
```

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 10]
```diff
@@ -179,6 +188,36 @@ KubernetesNamespace doCreateNamespaceAccess(String workspaceId, String name) {
         clientFactory, cheClientFactory, sharedPool.getExecutor(), name, workspaceId);
   }
 
+  @ScheduleRate(period = 1, initialDelay = 1, unit = TimeUnit.HOURS)
+  private void printLegacyWarningIfNeeded() {
+    if (!isNullOrEmpty(legacyNamespaceName)) {
+      LOG.warn(
+          "'che.infra.kubernetes.namespace' configuration parameter has been deprecated "
```
Comment (rkratky) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]:
```suggestion
          "The 'che.infra.kubernetes.namespace' configuration parameter has been deprecated "
```

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 11]
```diff
@@ -179,6 +188,36 @@ KubernetesNamespace doCreateNamespaceAccess(String workspaceId, String name) {
...
+              + "and is subject to removal in future releases.  Current value is: `{}`. Legacy workspaces located "
```
Comment (rkratky) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]:
```suggestion
              + "and is subject to removal in future releases. The current value is: `{}`. Legacy workspaces located "
```

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 12]
```diff
@@ -179,6 +188,36 @@ KubernetesNamespace doCreateNamespaceAccess(String workspaceId, String name) {
...
+              + "in this namespace may become unreachable in future releases. "
+              + "Please refer to the documentation about possible next steps.",
+          legacyNamespaceName);
+    }
+    if (LEGACY_NAMESPACE_NAME_PLACEHOLDERS.stream().anyMatch(defaultNamespaceName::contains)) {
+      LOG.warn(
+          "'che.infra.kubernetes.namespace.default' configuration parameter with `{}` "
```
Comment (rkratky) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]:
```suggestion
          "The 'che.infra.kubernetes.namespace.default' configuration parameter with the `{}` "
```

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 13]
```diff
@@ -179,6 +188,36 @@ KubernetesNamespace doCreateNamespaceAccess(String workspaceId, String name) {
...
+              + "placeholder has been deprecated and is subject to removal in future releases."
+              + " Current value is: `{}`. Please refer to the documentation about "
```
Comment (rkratky) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]:
```suggestion
              + " The current value is: `{}`. Please refer to the documentation about "
```

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 14]
```diff
@@ -179,6 +188,36 @@ KubernetesNamespace doCreateNamespaceAccess(String workspaceId, String name) {
...
+              + "possible next steps.",
+          LEGACY_NAMESPACE_NAME_PLACEHOLDERS,
+          defaultNamespaceName);
+    } else if (REQUIRED_NAMESPACE_NAME_PLACEHOLDERS
+        .stream()
+        .noneMatch(defaultNamespaceName::contains)) {
+      LOG.warn(
+          "Namespace strategies other than 'per user' have been deprecated and are subject to removal in future releases. "
+              + "Using the {} placeholder is required in 'che.infra.kubernetes.namespace.default' parameter."
```
Comment (rkratky) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]:
```suggestion
              + "Using the {} placeholder is required in the 'che.infra.kubernetes.namespace.default' parameter."
```

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 15]
```diff
@@ -179,6 +188,36 @@ KubernetesNamespace doCreateNamespaceAccess(String workspaceId, String name) {
...
+              + " Current value is: `{}`.",
```
Comment (rkratky) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]:
```suggestion
              + " The current value is: `{}`.",
```

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactoryTest.java [hunk 16]
```diff
@@ -1286,6 +1294,105 @@ public void normalizeLengthTest() {
...
+            NAMESPACE_LABELS,
+            NAMESPACE_ANNOTATIONS,
+            clientFactory,
+            cheClientFactory,
+            userManager,
+            preferenceManager,
+            pool);
+
+    assertEqualsMessage(
+        "'che.infra.kubernetes.namespace' configuration parameter has been deprecated and is subject to removal in future releases.  Current value is: `{}`. Legacy workspaces located in this namespace may become unreachable in future releases. Please refer to the documentation about possible next steps.");
```
Comment (rkratky) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactoryTest.java]:
```suggestion
        "The 'che.infra.kubernetes.namespace' configuration parameter has been deprecated and is subject to removal in future releases. The current value is: `{}`. Legacy workspaces located in this namespace may become unreachable in future releases. Please refer to the documentation about possible next steps.");
```

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactoryTest.java [hunk 17]
```diff
@@ -1286,6 +1294,105 @@ public void normalizeLengthTest() {
...
+            NAMESPACE_LABELS,
+            NAMESPACE_ANNOTATIONS,
+            clientFactory,
+            cheClientFactory,
+            userManager,
+            preferenceManager,
+            pool);
+
+    assertEqualsMessage(
+        "Namespace strategies other than 'per user' have been deprecated and are subject to removal in future releases. Using the {} placeholder is required in 'che.infra.kubernetes.namespace.default' parameter. Current value is: `{}`.");
```
Comment (rkratky) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactoryTest.java]:
```suggestion
        "Namespace strategies other than 'per user' have been deprecated and are subject to removal in future releases. Using the {} placeholder is required in the 'che.infra.kubernetes.namespace.default' parameter. The current value is: `{}`.");
```

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactoryTest.java [hunk 18]
```diff
@@ -1286,6 +1294,105 @@ public void normalizeLengthTest() {
...
+            NAMESPACE_LABELS,
+            NAMESPACE_ANNOTATIONS,
+            clientFactory,
+            cheClientFactory,
+            userManager,
+            preferenceManager,
+            pool);
+
+    assertEqualsMessage(
+        "'che.infra.kubernetes.namespace.default' configuration parameter with `{}` placeholder has been deprecated and is subject to removal in future releases. Current value is: `{}`. Please refer to the documentation about possible next steps.");
```
Comment (rkratky) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactoryTest.java]:
```suggestion
        "The 'che.infra.kubernetes.namespace.default' configuration parameter with the `{}` placeholder has been deprecated and is subject to removal in future releases. The current value is: `{}`. Please refer to the documentation about possible next steps.");
```

**Review by rkratky — APPROVED** [review]
@skabashnyuk, I added a bunch of language nitpicks and preapproved.

**Pull request merged**